pkg/auth/userregistry/identitymapper/strategy_deconflict.go

```diff
@@ -0,0 +1,86 @@
+package identitymapper
+
+import (
+	"errors"
+	"fmt"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	kerrs "k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/util/sets"
+
+	"github.com/openshift/origin/pkg/user"
+	userapi "github.com/openshift/origin/pkg/user/api"
+	userregistry "github.com/openshift/origin/pkg/user/registry/user"
+)
+
+// UserNameGenerator returns a username
+type UserNameGenerator func(base string, sequence int) string
+
+var (
+	// MaxGenerateAttempts limits how many times we try to find an available username for a new identity
+	MaxGenerateAttempts = 100
+
+	// DefaultGenerator attempts to use the base name first, then "base2", "base3", ...
+	DefaultGenerator = UserNameGenerator(func(base string, sequence int) string {
+		if sequence == 0 {
+			return base
+		}
+		return fmt.Sprintf("%s%d", base, sequence+1)
+	})
+)
+
+type StrategyDeconflict struct {
+	user        userregistry.Registry
+	generator   UserNameGenerator
+	initializer user.Initializer
+}
+
+func NewStrategyDeconflict(user userregistry.Registry, initializer user.Initializer) UserStrategy {
+	return &StrategyDeconflict{user, DefaultGenerator, initializer}
+}
+
+func (s *StrategyDeconflict) UserForIdentity(ctx kapi.Context, preferredUserName string, identity *userapi.Identity) (*userapi.User, error) {
+
+	// Iterate through the max allowed generated usernames
+	// If an existing user references this identity, associate the identity with that user and return
+	// Otherwise, create a user with the first generated user name that does not already exist and return.
+	// Names are created in a deterministic order, so the first one that isn't present gets created.
+	// In the case of a race, one will get to persist the user object and the other will fail.
+	for sequence := 0; sequence < MaxGenerateAttempts; sequence++ {
+		// Get the username we want
+		potentialUserName := s.generator(preferredUserName, sequence)
+
+		// See if it already exists
+		persistedUser, err := s.user.GetUser(ctx, potentialUserName)
+
+		if err != nil && !kerrs.IsNotFound(err) {
+			// Fail on errors other than "not found"
+			return nil, err
+		}
+
+		if err != nil && kerrs.IsNotFound(err) {
+			// Try to create a user with the available name
```

Comment:
Could change this code block to be parallel with the others, does the same thing. Shared method?